plugins/languages/treesitter/treesitter.nix

```
@@ -6,210 +6,316 @@
+        This is required to build grammars if you are not using `nixGrammars`.
+        To disable the installation of GCC, set this option to `null`.
+      '';
+    };
+
+    grammarPackages = mkOption {
+      type = with types; listOf package;
+      default = config.plugins.treesitter.package.passthru.allGrammars or [ ];
+      example = literalExpression "pkgs.vimPlugins.nvim-treesitter.passthru.allGrammars";
+      defaultText = literalExpression "config.plugins.treesitter.package.passthru.allGrammars";
+      description = "Grammar packages to install";
+    };
+
+    languageRegister = mkOption {
+      type = with types; attrsOf (coercedTo str toList (listOf str));
+      default = { };
+      example = {
+        cpp = "onelab";
+        python = [
+          "foo"
+          "bar"
+        ];
       };
+      description = ''
+        This is a wrapping of the `vim.treesitter.language.register` function.
+        Register specific parsers to one or several filetypes.
+        The keys are the parser names and the values are either one or several filetypes.
+      '';
+    };
 
-      nixvimInjections = mkEnableOption "nixvim specific injections, like lua highlighting in extraConfigLua";
+    nixGrammars = mkOption {
+      type = types.bool;
+      default = true;
+      example = false;
+      description = "Whether to automatically install grammar packages.";
+    };
+
+    nixvimInjections = mkOption {
+      type = types.bool;
+      default = true;
+      example = false;
+      description = ''Whether to enable Nixvim injections, such as recognising extraConfigLua as lua.'';
+    };
+
+    nodejsPackage = helpers.mkPackageOption {
+      name = "nodejs";
+      default = pkgs.nodejs;
+      defaultText = literalExpression "pkgs.nodejs";
+      example = literalExpression "pkgs.nodejs_22";
+      description = ''
+        Which package (if any) to be added as the nodejs package.
+        This is required to build grammars if you are not using `nixGrammars`.
+        To disable the installation of NodeJS, set this option to `null`.
+      '';
+    };
+
+    treesitterPackage = helpers.mkPackageOption {
+      name = "treesitter";
+      default = pkgs.tree-sitter;
+      defaultText = literalExpression "pkgs.treesitter";
+      description = ''
+        Which package (if any) to be added as the treesitter binary.
+        This is required to build grammars if you are not using `nixGrammars`.
+        To disable the installation of treesitter, set this option to `null`.
+      '';
     };
   };
 
-  config =
-    let
-      tsOptions = {
-        highlight = {
-          inherit (cfg) enable;
-          disable = if (cfg.disabledLanguages != [ ]) then cfg.disabledLanguages else null;
+  # NOTE: We call require('nvim-treesitter.configs').setup in extraConfigLua.
+  callSetup = false;
+  # NOTE: We dont want to try installing the provided package because we install it in extraPlugins
+  installPackage = false;
```

```suggestion
  # We call setup manually below.
  callSetup = false;
  # We install cfg.package manually so we can install grammars using it.
  installPackage = false;
```

Maybe something along these lines would be bit less verbose?

---

@khaneliman you're welcome to add back the `NOTE:` prefix if you prefer.